File: algorithms/mouse.py
import numpy as np
from lidar import scan
from path import Path
from visualizer import visualize_maze, visualize_maze_gui
from ros_simulator_copy import RosSimulator
import time


# Define a class mouse that will be used to represent the mouse in the maze
class Mouse:

    def __init__(self, x, y, known_walls, goal, known_paths, known_moves, mode='left_first'):
        self.x = x
        self.y = y
        self.direction = (1, 0) # (1, 0): +x, (0, 1): +y, (-1, 0): -x, (0, -1): -y
        self.known_walls = known_walls
        self.goal = goal
        self.moves = 0
        self.known_paths = known_paths
        self.mode = mode  # 'left_first' or 'right_first'
        self.sim = RosSimulator()

    def analyze_tof_distances(self, tof_front, tof_left, tof_right):
        """
        Returns booleans for specific ToF distance conditions.

        Args:
            tof_front (float): Front ToF sensor distance (in cm).
            tof_left (float): Left ToF sensor distance (in cm).
            tof_right (float): Right ToF sensor distance (in cm).

        Returns:
            Tuple[bool, bool, bool, bool]: 
                1. Front distance between 10 and 15 cm
                2. Front distance between 35 and 40 cm
                3. Left distance between 24.438 and 33.425 cm
                4. Right distance between 24.438 and 33.425 cm
        """
        front_close = 0.01 <= tof_front <= 0.23
        front_far = 0.35 <= tof_front <= 0.5
        left_in_range = 0.21438 <= tof_left <= 0.36425
        right_in_range = 0.21438 <= tof_right <= 0.36425

        return front_close, front_far, left_in_range, right_in_range

    def scan_walls(self):
        # use lidar for scanning the walls around the mouse
        # Define the walls as a list of tuples representing blocked paths. For example, ((4, 0), (4, 1)) represents a wall between squares (4, 0) and (4, 1).
        # found_walls = scan(self.x, self.y, self.direction)
        # for wall in found_walls:
        #     if wall not in self.known_walls or ((wall[1], wall[0]) not in self.known_walls):
        #         print("Found a new wall between {} and {}".format(wall[0], wall[1]))
        #         self.known_walls.append(wall)

        tof_front = self.sim.get_tof_front_distance()
        tof_left = self.sim.get_tof_left_distance()
        tof_right = self.sim.get_tof_right_distance()
        print(tof_front, tof_left, tof_right)
        front_close, front_far, left_in_range, right_in_range = self.analyze_tof_distances(tof_front, tof_left, tof_right)
        print(front_close, front_far, left_in_range, right_in_range )
        if front_close:
            wall = ((self.x, self.y), (self.x+self.direction[0], self.y+self.direction[1]))
            if wall not in self.known_walls:
                self.known_walls.append(wall)
            return
        
        if front_far:
            wall = ((self.x+self.direction[0], self.y+self.direction[1]), (self.x+2*self.direction[0], self.y+2*self.direction[1]))
            if wall not in self.known_walls:
                self.known_walls.append(wall)
        if left_in_range:
            wall = ((self.x+self.direction[0], self.y+self.direction[1]), (self.x+self.direction[0]+self.get_left()[0], self.y+self.direction[1]+self.get_left()[1]))
            if wall not in self.known_walls:
                self.known_walls.append(wall)
        if right_in_range:
            wall = ((self.x+self.direction[0], self.y+self.direction[1]), (self.x+self.direction[0]+self.get_right()[0], self.y+self.direction[1]+self.get_right()[1]))
            if wall not in self.known_walls:
                self.known_walls.append(wall)

        print("known walls: ", self.known_walls)

    
    def get_left(self):
        # Get the left direction based on the current direction
        if self.direction == (1, 0):
            return (0, 1)
        elif self.direction == (0, 1):
            return (-1, 0)
        elif self.direction == (-1, 0):
            return (0, -1)
        elif self.direction == (0, -1):
            return (1, 0)
        
    def get_right(self):
        # Get the right direction based on the current direction
        if self.direction == (1, 0):
            return (0, -1)
        elif self.direction == (0, 1):
            return (1, 0)
        elif self.direction == (-1, 0):
            return (0, 1)
        elif self.direction == (0, -1):
            return (-1, 0)
        
    def change_direction(self, new_direction):
        if new_direction == self.get_left():
            self.turn_left()
        elif new_direction == self.get_right():
            self.turn_right()
        elif new_direction[0] == -self.direction[0] and new_direction[1] == -self.direction[1]:
            self.turn_around()
        else:
            print("Invalid direction change requested: {}".format(new_direction))

    # interface functions
    def turn_left(self):
        self.sim.turn_left()
        self.direction = self.get_left()

    def turn_right(self):
        self.sim.turn_right()
        self.direction = self.get_right()

    def turn_around(self):
        self.sim.turn_left()
        self.sim.turn_left()
        self.turn_left()
        self.turn_left()

    def move_forward(self):
        feedback = self.sim.move_forward()
        position = feedback.pose.pose.position
        rel_x = (position.x - (-1.5))/0.25
        rel_y = (position.y - (-2.35))/0.25
        print(rel_x, rel_y)
        # self.x += self.direction[0]
        # self.y += self.direction[1]
        self.x = round(rel_x)
        self.y = round(rel_y)
        print("Moved to position ({}, {})".format(self.x, self.y))
        return True

    def flood_fill(self):
        maze = np.zeros((16, 16))
        l = maze.shape[0]
        b = maze.shape[1]
        
        # Initialize the queue with the goal position
        queue = [self.goal]
        
        # Process the queue until it's empty
        while queue: 
            # Get the current position from the queue
            x, y = queue.pop(0)
            
            # Check all four possible directions (right, left, down, up)
            for dx, dy in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
                # Skip if the move is blocked by a wall
                if ((x, y), (x + dx, y + dy)) in self.known_walls or ((x + dx, y + dy), (x, y)) in self.known_walls:
                    continue
                
                # Check if the new position is within the maze boundaries
                if 0 <= x + dx < l and 0 <= y + dy < b:
                    # If the new position is empty and not the goal, update its value and add to the queue
                    if maze[x + dx, y + dy] == 0 and (x + dx, y + dy) != self.goal:
                        maze[x + dx, y + dy] = maze[x, y] + 1
                        queue.append((x + dx, y + dy))
        
        # Return the updated maze
        return maze

    def navigate(self, maze, reverse=False):
        self.moves = 0
        l = maze.shape[0]
        b = maze.shape[1]
        print(maze.shape)

        moves = []
        known_walls_at_each_step = [self.known_walls.copy()] # for visualization purposes. Remove during production.
        positions = [(self.x, self.y)]
        while self.x != self.goal[0] or self.y != self.goal[1]:
            visualize_maze(self, maze)
            best_value = -1
            best_move = (0, 0)
            # Moves to the adjacent square with the least value
            if self.mode == 'left_first':
                # Check left, forward, right, backward in that order
                directions = [self.direction, self.get_left(), self.get_right(), (-self.direction[0], -self.direction[1])]
            else:
                # Check right, forward, left, backward in that order
                directions = [self.direction, self.get_right(), self.get_left(), (-self.direction[0], -self.direction[1])]
            for dx, dy in directions:
                    # Skip if the move is blocked by a wall
                    if ((self.x, self.y), (self.x + dx, self.y + dy)) in self.known_walls or ((self.x + dx, self.y + dy), (self.x, self.y)) in self.known_walls:
                        continue
                    
                    # Check if the new position is within the maze boundaries
                    if 0 <= self.x + dx < l and 0 <= self.y + dy < b:
                        if maze[self.x + dx, self.y + dy] < best_value or best_value == -1:
                            best_value = maze[self.x + dx, self.y + dy]
                            best_move = (dx, dy)
            # Move the mouse to the best position
            if self.direction == (best_move[0], best_move[1]):
                # If the best move is in the same direction, move forward
                self.move_forward()
                self.moves += 1
                moves += [best_move]
                positions += [(self.x, self.y)]
                known_walls_at_each_step.append(self.known_walls.copy())
            else:
                # If the best move is in a different direction, change direction
                self.change_direction((best_move[0], best_move[1]))
                print("Changed direction to ({}, {})".format(self.direction[0], self.direction[1]))

            self.scan_walls()
            maze = self.flood_fill()

        visualize_maze(self, maze)
        visualize_maze_gui(self, maze, positions, known_walls_at_each_step)
        if reverse:
            # print("moves in reverse: ", moves)
            moves, positions = self.reverse(moves, positions)
        self.remove_loops_and_memorize(moves, positions)
        print("Goal reached in {} moves".format(self.moves))
        print("Number of moves after loop removal and optimization: {}".format(self.known_paths[-1].optimized_length))
        print("Number of turns after loop removal and optimization: {}".format(self.known_paths[-1].optimized_turns))
        print("Path: {}".format(self.known_paths[-1].positions))
        return
    
    def remove_loops_and_memorize(self, moves, positions):
        # removes the moves between instances when the same position is reached
        for i in range(len(moves)):
            for j in range(len(moves), i, -1):
                if positions[i] == positions[j]:
                    del moves[i:j]
                    del positions[i:j]
                    break
        print(len(moves), len(positions))
        self.known_paths.append(Path(positions, moves))
        return
    
    def reverse(self, moves, positions):
        # given the path taken from goal to starting point, determines the path from starting point to goal
        for i in range(len(moves)):
            moves[i] = (-moves[i][0], -moves[i][1])
        moves.reverse()
        positions.reverse() 
        return moves, positions
    
    def get_best_path(self):
        # Return the path with the highest feasibility score
        best_path = None
        for path in self.known_paths:
            if best_path is None or path.feasibility_score > best_path.feasibility_score:
                best_path = path
        return best_path
    
    


    

File: algorithms_cpp/Mouse.cpp
#include "Mouse.hpp"
#include <algorithm>
#include <cmath>
#include <iostream>
#include <queue>
#include <unordered_map>
#include <set>
#include <memory>
#include "Lidar.hpp"

Mouse::Mouse(int x, int y,
             const std::set<std::pair<std::pair<int, int>, std::pair<int, int>>>& known_walls,
             std::pair<int, int> goal,
             const std::vector<std::vector<int>>& maze,
             const std::vector<Path>& known_paths,
             const std::string& mode)
    : x_(x), y_(y), knownWalls_(known_walls), goal_(goal),
      maze_(maze), knownPaths_(known_paths), mode_(mode),
      direction_({0, 1}) {}  // assuming default facing North

void Mouse::setGoal(const std::pair<int, int>& newGoal) {
    goal_ = newGoal;
}

std::vector<Path> Mouse::getKnownPaths() const {
    return knownPaths_;
}

std::tuple<bool, bool, bool, bool> Mouse::analyzeTofDistances(float front, float left, float right) {
    bool frontClose = 0.01 <= front && front <= 0.23;
    bool frontFar = 0.35 <= front && front <= 0.5;
    bool leftInRange = (0.2316 - 0.075) <= left && left <= (0.2316 + 0.075);
    bool rightInRange = (0.2316 - 0.075) <= right && right <= (0.2316 + 0.075);
    return std::make_tuple(frontClose, frontFar, leftInRange, rightInRange);
}

std::pair<int, int> Mouse::getLeft() {
    if (direction_ == std::make_pair(1, 0)) return {0, 1};
    if (direction_ == std::make_pair(0, 1)) return {-1, 0};
    if (direction_ == std::make_pair(-1, 0)) return {0, -1};
    return {1, 0}; // direction_ == (0, -1)
}

std::pair<int, int> Mouse::getRight() {
    if (direction_ == std::make_pair(1, 0)) return {0, -1};
    if (direction_ == std::make_pair(0, 1)) return {1, 0};
    if (direction_ == std::make_pair(-1, 0)) return {0, 1};
    return {-1, 0}; // direction_ == (0, -1)
}

void Mouse::scanWalls() {
    auto found_walls = scan(x_, y_, direction_);
    for (const auto& wall : found_walls) {
        auto reverse_wall = std::make_pair(wall.second, wall.first);
        if (knownWalls_.find(wall) == knownWalls_.end() &&
            knownWalls_.find(reverse_wall) == knownWalls_.end()) {
            knownWalls_.insert(wall);
        }
    }
}

// void Mouse::scanWalls() {
//     float tofFront = sim_.getTofFrontDistance();
//     float tofLeft = sim_.getTofLeftDistance();
//     float tofRight = sim_.getTofRightDistance();
//     std::cout << "ToF readings: " << tofFront << ", " << tofLeft << ", " << tofRight << std::endl;

//     bool frontClose, frontFar, leftInRange, rightInRange;
//     std::tie(frontClose, frontFar, leftInRange, rightInRange) = analyzeTofDistances(tofFront, tofLeft, tofRight);

//     if (frontClose) {
//         calibrateFront();
//         auto wall = std::make_pair(std::make_pair(x_, y_),
//                                    std::make_pair(x_ + direction_.first, y_ + direction_.second));
//         if (std::find(knownWalls_.begin(), knownWalls_.end(), wall) == knownWalls_.end()) {
//             knownWalls_.push_back(wall);
//         }
//         return;
//     }

//     if (frontFar) {
//         auto wall = std::make_pair(
//             std::make_pair(x_ + direction_.first, y_ + direction_.second),
//             std::make_pair(x_ + 2 * direction_.first, y_ + 2 * direction_.second));
//         if (std::find(knownWalls_.begin(), knownWalls_.end(), wall) == knownWalls_.end()) {
//             knownWalls_.push_back(wall);
//         }
//     }

//     if (leftInRange) {
//         auto left = getLeft();
//         auto wall = std::make_pair(
//             std::make_pair(x_ + direction_.first, y_ + direction_.second),
//             std::make_pair(x_ + direction_.first + left.first, y_ + direction_.second + left.second));
//         if (std::find(knownWalls_.begin(), knownWalls_.end(), wall) == knownWalls_.end()) {
//             knownWalls_.push_back(wall);
//         }
//     }

//     if (rightInRange) {
//         auto right = getRight();
//         auto wall = std::make_pair(
//             std::make_pair(x_ + direction_.first, y_ + direction_.second),
//             std::make_pair(x_ + direction_.first + right.first, y_ + direction_.second + right.second));
//         if (std::find(knownWalls_.begin(), knownWalls_.end(), wall) == knownWalls_.end()) {
//             knownWalls_.push_back(wall);
//         }
//     }

//     std::cout << "Known walls: " << knownWalls_.size() << std::endl;
// }


// Utility movement methods
void Mouse::turnLeft() {
    direction_ = {-direction_.second, direction_.first};
    sim_.turnLeft();
}

void Mouse::turnRight() {
    direction_ = {direction_.second, -direction_.first};
    sim_.turnRight();
}

void Mouse::turnAround() {
    direction_ = {-direction_.first, -direction_.second};
    sim_.turnRight();
    sim_.turnRight();
}

void Mouse::changeDirection(std::pair<int, int> newDir) {
    if (newDir == direction_) return;
    if (newDir == std::make_pair(-direction_.second, direction_.first)) {
        turnLeft();
    } else if (newDir == std::make_pair(direction_.second, -direction_.first)) {
        turnRight();
    } else {
        turnAround();
    }
}

bool Mouse::moveForward(double distance) {
    // Feedback feedback = sim_.moveForward(distance);
    // const auto& position = feedback.pose.pose.position;

    // double relX = (position.x) / 0.25;
    // double relY = (position.y) / 0.25;

    // std::cout << relX << ", " << relY << std::endl;

    x_ = x_ + direction_.first;
    y_ = y_ + direction_.second;

    std::cout << "Moved to position (" << x_ << ", " << y_ << ")" << std::endl;

    // correctPosition(relX, relY);
    return true;
}

void Mouse::moveForward() {
    moveForward(1.0); // Move forward by one cell (or suitable default)
}

// void Mouse::calibrateFront() {
//     float target = 0.094;
//     float tof = sim_.getTofFrontDistance();
//     float correction = tof - target;
//     float velocity = (correction < 0) ? -0.1f : 0.1f;
//     sim_.moveForward(std::abs(correction), velocity);
//     std::cout << "Calibrated front distance with correction: " << correction << std::endl;
// }

std::unordered_map<std::pair<int, int>, int, pair_hash> Mouse::floodFill(
    std::pair<int, int> goal,
    const std::vector<std::pair<std::pair<int, int>, std::pair<int, int>>>& walls,
    const std::set<std::pair<int, int>>& visited) {
    std::unordered_map<std::pair<int, int>, int, pair_hash> distances;
    std::queue<std::pair<int, int>> q;
    q.push(goal);
    distances[goal] = 0;

    auto isWall = [&](const std::pair<int, int>& from, const std::pair<int, int>& to) {
        return std::find(walls.begin(), walls.end(), std::make_pair(from, to)) != walls.end() ||
               std::find(walls.begin(), walls.end(), std::make_pair(to, from)) != walls.end();
    };

    std::vector<std::pair<int, int>> directions = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

    while (!q.empty()) {
        auto curr = q.front();
        q.pop();
        int currDist = distances[curr];

        for (const auto& dir : directions) {
            auto next = std::make_pair(curr.first + dir.first, curr.second + dir.second);

            if (visited.find(next) != visited.end() &&
                distances.find(next) == distances.end() &&
                !isWall(curr, next)) {
                distances[next] = currDist + 1;
                q.push(next);
            }
        }
    }

    return distances;
}

std::vector<std::vector<int>> Mouse::floodFill() {
    // Prepare visited set
    std::set<std::pair<int, int>> visited;
    int l = maze_.size();
    int b = maze_[0].size();
    for (int i = 0; i < l; ++i)
        for (int j = 0; j < b; ++j)
            visited.insert({i, j});

    // Convert knownWalls_ to vector
    std::vector<std::pair<std::pair<int, int>, std::pair<int, int>>> walls(knownWalls_.begin(), knownWalls_.end());

    // Get distances from floodFill
    auto distances = floodFill(goal_, walls, visited);

    // Build maze grid from distances
    std::vector<std::vector<int>> maze(l, std::vector<int>(b, std::numeric_limits<int>::max()));
    for (const auto& [pos, dist] : distances) {
        maze[pos.first][pos.second] = dist;
    }
    return maze;
}


void Mouse::navigate(const std::vector<std::vector<int>>& maze, bool reverse) {
    moves_ = 0;
    int l = maze.size();
    int b = maze[0].size();

    std::vector<std::pair<int, int>> moves;
    std::vector<std::set<std::pair<std::pair<int, int>, std::pair<int, int>>>> known_walls_at_each_step;
    known_walls_at_each_step.push_back(knownWalls_);  // For visualization

    std::vector<std::pair<int, int>> positions;
    positions.emplace_back(x_, y_);

    while (x_ != goal_.first || y_ != goal_.second) {
        visualizer_.visualizeMaze(maze_, x_, y_, knownWalls_);
        int best_value = std::numeric_limits<int>::max();
        std::pair<int, int> best_move = {0, 0};

        std::vector<std::pair<int, int>> directions;
        if (mode_ == "left_first") {
            directions = {direction_, getLeft(), getRight(), {-direction_.first, -direction_.second}};
        } else {
            directions = {direction_, getRight(), getLeft(), {-direction_.first, -direction_.second}};
        }

        for (auto [dx, dy] : directions) {
            int new_x = x_ + dx;
            int new_y = y_ + dy;

            if (knownWalls_.count({{x_, y_}, {new_x, new_y}}) || knownWalls_.count({{new_x, new_y}, {x_, y_}}))
                continue;

            if (new_x >= 0 && new_x < l && new_y >= 0 && new_y < b) {
                int value = maze_[new_x][new_y];
                if (value < best_value) {
                    best_value = value;
                    best_move = {dx, dy};
                }
            }
        }

        if (direction_ == best_move) {
            moveForward();
            moves_++;
            moves.push_back(best_move);
            positions.emplace_back(x_, y_);
            known_walls_at_each_step.push_back(knownWalls_);
        } else {
            changeDirection(best_move);
        }

        scanWalls();
        maze_ = floodFill();
    }

    visualizer_.visualizeMaze(maze_, x_, y_, knownWalls_);

    if (reverse) {
        std::tie(moves, positions) = reversePath(moves, positions);
    }

    removeLoopsAndMemorize(moves, positions);

    std::cout << "Goal reached in " << moves_ << " moves\n";
    std::cout << "Number of moves after loop removal and optimization: " << knownPaths_.back().optimizedLength << "\n";
    std::cout << "Number of turns after loop removal and optimization: " << knownPaths_.back().optimizedTurns << "\n";
    std::cout << "Path: ";
    for (auto& pos : knownPaths_.back().positions) {
        std::cout << "(" << pos.first << "," << pos.second << ") ";
    }
    std::cout << "\n";
}

void Mouse::removeLoopsAndMemorize(const std::vector<std::pair<int, int>>& moves, 
                                std::vector<std::pair<int, int>>& path) {
    std::unordered_map<std::pair<int, int>, int, pair_hash> seen;
    std::vector<std::pair<int, int>> newPath;

    for (size_t i = 0; i < path.size(); ++i) {
        auto cell = path[i];
        if (seen.find(cell) != seen.end()) {
            newPath.erase(newPath.begin() + seen[cell], newPath.end());
        } else {
            seen[cell] = newPath.size();
            newPath.push_back(cell);
        }
    }

    Path newPathObj(newPath, moves);
    knownPaths_.push_back(newPathObj);
}

std::pair<std::vector<std::pair<int, int>>, std::vector<std::pair<int, int>>>
Mouse::reversePath(std::vector<std::pair<int, int>> moves,
                   std::vector<std::pair<int, int>> positions) {
    for (auto& move : moves) {
        move.first = -move.first;
        move.second = -move.second;
    }
    std::reverse(moves.begin(), moves.end());
    std::reverse(positions.begin(), positions.end());

    return {moves, positions};
}

std::shared_ptr<Path> Mouse::getBestPath() const {
    std::shared_ptr<Path> bestPath = nullptr;
    for (const auto& path : knownPaths_) {
        if (!bestPath || path.feasibilityScore > bestPath->feasibilityScore) {
            bestPath = std::make_shared<Path>(path);
        }
    }
    return bestPath;
}
